src/miniscript/satisfy.rs

```diff
@@ -212,339 +238,521 @@ impl_tuple_satisfier!(&'a A, &'b B, &'c C, &'d D, &'e E, &'f F,);
-                    return None;
-                }
-                if satisfied == k {
-                    return Some(flatten(ret));
-                }
-
-                // If we have more satisfactions than needed, throw away the
-                // extras, choosing the ones that yield the biggest savings.
-                let mut indices: Vec<usize> = (0..subs.len()).collect();
-                indices.sort_by(|a, b| {
-                    if !subs[*a].ty.mall.safe && subs[*b].ty.mall.safe {
-                        Ordering::Less
-                    } else if subs[*a].ty.mall.safe && !subs[*b].ty.mall.safe {
-                        Ordering::Greater
-                    } else {
-                        let a_cost =
-                            satisfy_cost(&ret_dis[*a]) as isize - satisfy_cost(&ret[*a]) as isize;
-                        let b_cost =
-                            satisfy_cost(&ret_dis[*b]) as isize - satisfy_cost(&ret[*b]) as isize;
-                        if a_cost < b_cost {
-                            Ordering::Less
-                        } else {
-                            Ordering::Greater
-                        }
+                } else {
+                    // Otherwise flatten everything out
+                    Satisfaction {
+                        has_sig: ret_stack.iter().any(|sat| sat.has_sig),
+                        stack: ret_stack.into_iter().fold(
+                            Witness::empty(),
+                            |acc, next| Witness::combine(next.stack, acc),
+                        ),
                     }
-                });
-                for i in indices.iter().take(satisfied - k) {
-                    mem::swap(&mut ret[*i], &mut ret_dis[*i]);
                 }
-
-                Some(flatten(ret))
-            }
+            },
             Terminal::ThreshM(k, ref keys) => {
-                let mut ret = Vec::with_capacity(k + 1);
-
-                ret.push(vec![]);
+                // Collect all available signatures
+                let mut sig_count = 0;
+                let mut sigs = Vec::with_capacity(k);
                 for pk in keys {
-                    if let Some(vec) = satisfier.lookup_pk_vec(pk) {
-                        ret.push(vec);
-                        if ret.len() > k + 1 {
-                            let max_idx = ret
-                                .iter()
-                                .enumerate()
-                                .max_by_key(|&(_, ref sig)| sig.len())
-                                .unwrap()
-                                .0;
-                            ret.remove(max_idx);
+                    match Witness::signature(stfr, pk) {
+                        Witness::Stack(sig) => {
+                            sigs.push(sig);
+                            sig_count += 1;
+                        },
+                        Witness::Unavailable => {
+                            sigs.push(vec![]);
```

Comment:
Unavailable should not be push anything onto the stack.

Reply:
There seems to something wrong in the satisfaction for multi. I think you may have forgotten to remove the vec![] from the sigs vector

Reply:
Oh, you're right, I had a broken model of CHECKMULTISIG in my head.